src/NHibernate/Impl/FutureBatch.cs

```diff
@@ -80,5 +116,88 @@ private IEnumerable<TResult> GetCurrentResult<TResult>(int currentIndex)
 		protected abstract void ClearCurrentFutureBatch();
 		protected abstract bool IsQueryCacheable(TQueryApproach query);
 		protected abstract string CacheRegion(TQueryApproach query);
+
+		protected virtual void AddResultTransformer(
+			TMultiApproach multiApproach,
+			IResultTransformer futureResulsTransformer)
+		{
+			// Only Linq set ExecuteOnEval, so only FutureQueryBatch needs to support it, not FutureCriteriaBatch.
+			throw new NotSupportedException();
+		}
+
+		[Serializable]
+		private class BatchedQueryPostExecute
+		{
+			public System.Type ResultType { get; set; }
+			public Delegate ExecuteOnEval { get; set; }
+			public bool IsValue { get; set; }
+		}
+
+		// ResultTransformer are usually re-usable, this is not the case of this one, which will
+		// be built for each multi-query requiring it.
+		// It also usually ends in query cache, but this is not the case either for multi-query.
+		[Serializable]
+		private class FutureResultsTransformer : IResultTransformer
```

Comment:
Since `FutureResultsTransformer` isn't actually serializable because `BatchedQuery` isn't serializable, can the test in `DetachedCriteriaSerializable` be adjusted to exclude this type instead?  `DetachedCriteria` doesn't expose a `Future`, so not all `IResultTransformer` technically need to be serializable.

Why is serializing a `DetachedCriteria` so important anyway?

Reply:
Originally I have made that class truly serializable, but that was requiring an additional "data holding" class just for that.

We could instead try to adjust the serializable test, but since this class is private, we will have to add some more reflection logic matching by name to do so. Does that `Serializable` attribute hurts somewhere that it would be better to remove it?

About `DetachedCriteria`, I have not checked why it actually needs to be serializable, but I think it is likely because it may ends in the query cache too.

Reply:
Since serialization is the biggest compatibility break with .NET Core, I am writing some Roslyn analyzers to find issues so I can try to resolve them ahead of time.  On types that are required to be serialized for some reason, and use `Type` or `MethodInfo`, etc., I'll probably have to implement `ISerializable`.

Of course this `FutureResultsTransformer ` is triggering a full-framework warning which is based on FxCop rule [CA2235](https://docs.microsoft.com/visualstudio/code-quality/ca2235-mark-all-non-serializable-fields).  Marking the field as `[NonSerializable]` is really just masking the problem because the entire type wouldn't work if it were deserialized like that.

Reply:
It seems to me most of those `Serializable` things everywhere in NHibernate code are here for allowing to serialize sessions and session factories. But is there any practical common use case in serializing sessions? Do a significant parts of NHibernate user need this?

Maybe should we consider dropping the possibility of serializing sessions and session factories.

Reply:
I found [this](https://nhibernate.jira.com/browse/NH-3487), which shows there are users using session serialization. It looks like we need to keep that.